delfin/drivers/dell_emc/vnx_block/navi_handler.py

```diff
@@ -0,0 +1,469 @@
+from delfin.drivers.utils.tools import Tools
+
+LOG = logging.getLogger(__name__)
+
+
+class NaviHandler(object):
+    """Common class for EMC VNX storage system."""
+
+    USER_SECURITY_API = 'naviseccli -AddUserSecurity -User %(username)s ' \
+                        '-password %(password)s -h %(host)s -scope 0 ' \
+                        '-t %(timeout)d'
+    REMOVEUSERSECURITY_API = 'naviseccli -h %(host)s -RemoveUserSecurity'
+    GET_AGENT_API = 'naviseccli -h %(host)s getagent'
+    GET_DOMAIN_API = 'naviseccli -h %(host)s domain -list'
+    GET_STORAGEPOOL_API = 'naviseccli -h %(host)s storagepool -list'
+    GET_RAIDGROUP_API = 'naviseccli -h %(host)s getall -rg'
+    GET_DISK_API = 'naviseccli -h %(host)s getdisk'
+    GET_LUN_API = 'naviseccli -h %(host)s lun -list'
+    GET_GETALLLUN_API = 'naviseccli -h %(host)s getall -lun'
+    GET_LOG_API = 'naviseccli -h %(host)s getlog ' \
+                  '-date %(begin_time)s %(end_time)s'
+
+    TIME_PATTERN = '%m/%d/%Y %H:%M:%S'
+    DATE_PATTERN = '%m/%d/%Y'
+    ONE_DAY_SCE = 24 * 60 * 60
+    SOCKET_TIMEOUT = 5
+
+    def __init__(self, **kwargs):
+        ssh_access = kwargs.get('ssh')
+        if ssh_access is None:
+            raise exception.InvalidInput('Input navicli_access is missing')
+        self.navi_host = ssh_access.get('host')
+        self.navi_port = ssh_access.get('port')
+        self.navi_username = ssh_access.get('username')
+        self.navi_password = cryptor.decode(ssh_access.get('password'))
+        self.navi_timeout = ssh_access.get('conn_timeout')
+        if self.navi_timeout is None:
+            self.navi_timeout = NaviClient.SOCKET_TIMEOUT
+
+    def login(self, host_ip=None):
+        """Test connection """
+        version = ''
+        try:
+            if host_ip is None:
+                host_ip = self.navi_host
+            navi_client = NaviClient()
+            command_str = self.USER_SECURITY_API % {
+                'username': self.navi_username, 'password': self.navi_password,
+                'host': host_ip, 'timeout': self.navi_timeout}
+            if self.navi_port:
+                command_str = '%s -port %d' % (command_str, self.navi_port)
+            navi_client.exec(command_str.split())
+            command_str = self.GET_AGENT_API % {'host': host_ip}
+            result = navi_client.exec(command_str.split(), stdin_value='2\r\n')
+            if result:
+                agent_model = self.arrange_resource_obj(result)
+                if agent_model:
+                    version = agent_model.get("revision")
+        except exception.NaviCliConnectTimeout as e:
+            raise e
+        except exception.InvalidUsernameOrPassword as e:
+            raise e
+        except Exception as e:
+            err_msg = "Login error: %s" % (six.text_type(e))
+            raise exception.InvalidResults(err_msg)
+        return version
+
+    def get_agent(self):
+        """get agent info"""
+        agent_model = {}
+        try:
```

Comment:
How about moving the whole try catch block into common function and here simply call that common function. This common function can return result

Reply:
Implementation method self.navi_ Exe is a public method, but different resource contents return different processing.

Reply:
Yes, i got that..i was saying that there is lots of code repetition which can be avoided by simply getting the result and processing separately

Reply:
Modified